frostsnapp/native/src/coordinator.rs

```diff
@@ -31,55 +36,88 @@ impl FfiCoordinator {
-                        device_changes,
-                        new_messages,
-                    } = { loop_manager.lock().unwrap().poll_ports() };
-
-                    if !device_changes.is_empty() {
-                        crate::api::emit_device_events(
-                            device_changes
-                                .into_iter()
-                                .map(crate::api::DeviceChange::from)
-                                .collect(),
+                new_messages
+            };
+
+            for (from, message) in new_messages {
+                match coordinator_loop
+                    .lock()
+                    .unwrap()
+                    .recv_device_message(from, message.clone())
+                {
+                    Ok(messages) => {
+                        let mut pending_messages = pending_loop.lock().unwrap();
+                        pending_messages.extend(messages);
+                    }
+                    Err(e) => {
+                        event!(
+                            Level::ERROR,
+                            from = from.to_string(),
+                            "Failed to process message: {}",
+                            e
                         );
+                        continue;
                     }
-
-                    new_messages
                 };
+            }
 
-                for (from, message) in new_messages {
-                    match core_coordinator.recv_device_message(from, message.clone()) {
-                        Ok(messages) => {
-                            outbox.extend(messages);
-                        }
-                        Err(e) => {
-                            event!(
-                                Level::ERROR,
-                                from = from.to_string(),
-                                "Failed to process message: {}",
-                                e
-                            );
-                            continue;
+            let mut pending_messages = pending_loop.lock().unwrap();
+            while let Some(message) = pending_messages.pop_front() {
+                match message {
+                    CoordinatorSend::ToDevice(msg) => {
+                        let send_message = CoordinatorSendMessage {
+                            target_destinations: msg.default_destinations(),
+                            message_body: CoordinatorSendBody::Core(msg),
+                        };
+
+                        manager_thread
+                            .lock()
+                            .unwrap()
+                            .queue_in_port_outbox(send_message);
+                    }
+                    CoordinatorSend::ToUser(msg) => match msg {
+                        CoordinatorToUserMessage::Signed { .. } => {}
+                        CoordinatorToUserMessage::CheckKeyGen { .. } => {
+                            // Don't want to ack this until button press!!!
```

Comment:
I had this previously but in did it, since it was superfluous to the keygen progression UI.

This is because the keygen thread polling for a `FrostKey` state would finish at the same time as `bool: coordinator.is_awaiting_keygen` became true. So it was doubling up and had the feeling that I was adding too many state variables to the coordinator api. But I'll add it back and try with other UX changes

Reply:
I think this all works nicely now.